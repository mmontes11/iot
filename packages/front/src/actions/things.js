import {
  THINGS_UPDATED,
  THING_SELECTED,
  THINGS_REQUEST,
  THINGS_REQUEST_SUCCESS,
  THINGS_REQUEST_ERROR,
  SHOW_THING_NOT_FOUND_ERROR,
} from "constants/actionTypes/things";
import iotClient from "lib/iotClient";

export const getThings = () => dispatch => {
  dispatch({ type: THINGS_REQUEST });
  iotClient.thingsService
    .getThings()
    .then(res => {
      dispatch({ type: THINGS_REQUEST_SUCCESS, statusCode: res.statusCode, error: null });
      dispatch({ type: THINGS_UPDATED, things: res.body.things });
    })
    .catch(error => dispatch({ type: THINGS_REQUEST_ERROR, statusCode: error.statusCode, error }));
};

export const selectThing = thing => dispatch => {
  dispatch({ type: THING_SELECTED, thing });
};

export const showThingNotFoundError = showNotFoundError => dispatch => {
  dispatch({ type: SHOW_THING_NOT_FOUND_ERROR, showNotFoundError });
};
