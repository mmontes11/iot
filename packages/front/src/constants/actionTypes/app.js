export const TOGGLE_HAMBURGER_MENU = "TOGGLE_HAMBURGER_MENU";
export const TOGGLE_MAP_DIALOG = "TOGGLE_MAP_DIALOG";
export const SET_SHOW_ERROR = "SET_SHOW_ERROR";
