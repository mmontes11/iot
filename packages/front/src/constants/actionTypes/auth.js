export const IS_AUTH = "IS_AUTH";
export const SET_USERNAME = "SET_USERNAME";
export const SET_PASSWORD = "SET_PASSWORD";
export const LOGIN_REQUEST = "LOGIN_REQUEST";
export const LOGIN_REQUEST_SUCCESS = "LOGIN_REQUEST_SUCCESS";
export const LOGIN_REQUEST_ERROR = "LOGIN_REQUEST_ERROR";
