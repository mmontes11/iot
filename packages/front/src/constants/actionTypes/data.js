export const DATA_UPDATED = "DATA_UPDATED";
export const DATA_REQUEST = "DATA_REQUEST";
export const DATA_REQUEST_SUCCESS = "DATA_REQUEST_SUCCESS";
export const DATA_REQUEST_ERROR = "DATA_REQUEST_ERROR";
export const ADD_DATA_ITEM = "ADD_DATA_ITEM";
export const RESET_DATA = "RESET_DATA";
