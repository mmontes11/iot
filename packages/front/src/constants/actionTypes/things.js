export const THINGS_UPDATED = "THINGS_UPDATED";
export const THING_SELECTED = "THING_SELECTED";
export const THINGS_REQUEST = "THINGS_REQUEST";
export const THINGS_REQUEST_SUCCESS = "THINGS_REQUEST_SUCCESS";
export const THINGS_REQUEST_ERROR = "THINGS_REQUEST_ERROR";
export const SHOW_THING_NOT_FOUND_ERROR = "SHOW_THING_NOT_FOUND_ERROR";
