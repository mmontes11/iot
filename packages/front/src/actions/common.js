import { RESET } from "constants/actionTypes/common";

export const reset = preserveError => dispatch => {
  dispatch({ type: RESET, preserveError });
};
