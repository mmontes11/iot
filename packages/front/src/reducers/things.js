import { THINGS_UPDATED, THING_SELECTED, SHOW_THING_NOT_FOUND_ERROR } from "constants/actionTypes/things";
import { RESET } from "constants/actionTypes/common";

export const initialState = {
  items: [],
  selectedItem: null,
  showNotFoundError: false,
};

export default (state = initialState, { type, things, thing, showNotFoundError }) => {
  switch (type) {
    case THINGS_UPDATED:
      return { ...state, items: things };
    case THING_SELECTED:
      if (state.selectedItem !== null && state.selectedItem.name === thing.name) {
        return { ...state, selectedItem: null };
      }
      return { ...state, selectedItem: thing };
    case SHOW_THING_NOT_FOUND_ERROR:
      return { ...state, showNotFoundError };
    case RESET:
      return { ...initialState, items: state.items };
    default:
      return state;
  }
};
