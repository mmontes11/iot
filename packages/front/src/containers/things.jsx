import React from "react";
import { connect } from "react-redux";
import { withRouter } from "react-router-dom";
import { compose } from "recompose";
import PropTypes from "prop-types";
import ThingItem from "components/thingItem";
import ThingDetail from "components/thingDetail";
import Loader from "components/loader";
import Modal from "components/modal";
import * as thingActions from "actions/things";
import * as fromState from "reducers";
import { withResetOnUnmount } from "hocs/resetOnUnmount";
import { injectIntl, intlShape } from "react-intl";
import { EVENT_TYPE, MEASUREMENT_TYPE } from "constants/observationTypes";
import { defaultGroupBy } from "config/params";

class Things extends React.Component {
  componentDidMount() {
    if (this.props.things.length === 0) {
      this.props.getThings();
    }
  }
  componentDidUpdate() {
    const {
      selectedThing,
      things,
      match: {
        params: { thing: thingName },
      },
    } = this.props;
    if (selectedThing === null && things.length > 0 && thingName) {
      const thing = things.find(t => t.name === thingName);
      if (thing) {
        this._selectThing(thing);
      } else {
        this._resetPathToRoot();
        this.props.showNotFoundError(true);
      }
    }
  }
  _isSelected = thing => this.props.selectedThing !== null && this.props.selectedThing.name === thing.name;
  _selectThing = thing => {
    this.props.selectThing(thing);
    if (this._isSelected(thing)) {
      this._resetPathToRoot();
    } else {
      this.props.history.push(`/things/${thing.name}`);
    }
  };
  _resetPathToRoot = () => {
    this.props.history.push("/things");
  };
  _onStatsClick = (type, thing) => {
    const supportedEvents = thing.supportedObservationTypes.event;
    const supportedMeasurements = thing.supportedObservationTypes.measurement;
    let url;
    if (type === EVENT_TYPE && supportedEvents && supportedEvents.length > 0) {
      url = `/stats/${type}/${supportedEvents[0]}?thing=${thing.name}&timePeriod=day`;
    } else if (type === MEASUREMENT_TYPE && supportedMeasurements && supportedMeasurements.length > 0) {
      url = `/stats/${type}/${supportedMeasurements[0]}?thing=${thing.name}&timePeriod=day`;
    }
    if (url) {
      this.props.history.push(url);
    }
  };
  _onDataClick = (type, observation, thing) => {
    this.props.history.push(`/data/${type}/${observation}?thing=${thing.name}&groupBy=${defaultGroupBy}`);
  };
  render() {
    const {
      intl: { formatMessage },
      isLoading,
      things,
      selectedThing,
      shouldShowNotFoundError,
      showNotFoundError,
    } = this.props;
    if (isLoading && things.length === 0) {
      return <Loader />;
    }
    return (
      <div>
        <div className="container is-fluid section">
          <div className="columns">
            <div className="column is-one-quarter">
              {things.map(thing => (
                <ThingItem
                  key={thing.name}
                  name={thing.name}
                  isSelected={this._isSelected(thing)}
                  onClick={() => this._selectThing(thing)}
                />
              ))}
            </div>
            {selectedThing && (
              <div className="column is-three-quarters">
                <ThingDetail
                  thing={selectedThing}
                  onStatsClick={type => this._onStatsClick(type, selectedThing)}
                  onDataClick={(type, observation) => this._onDataClick(type, observation, selectedThing)}
                />
              </div>
            )}
          </div>
        </div>
        <Modal
          isActive={shouldShowNotFoundError}
          onCloseClick={() => showNotFoundError(false)}
          messageStyle="is-danger"
          title={formatMessage({ id: "Error" })}
          subTitle={formatMessage({ id: "Thing not found" })}
        />
      </div>
    );
  }
}

Things.propTypes = {
  intl: intlShape.isRequired,
  getThings: PropTypes.func.isRequired,
  selectThing: PropTypes.func.isRequired,
  selectedThing: PropTypes.shape({
    name: PropTypes.string.isRequired,
  }),
  isLoading: PropTypes.bool.isRequired,
  things: PropTypes.arrayOf(PropTypes.shape({})),
  shouldShowNotFoundError: PropTypes.bool,
  showNotFoundError: PropTypes.func.isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }).isRequired,
  match: PropTypes.shape({}).isRequired,
};

Things.defaultProps = {
  selectedThing: null,
  things: [],
  shouldShowNotFoundError: false,
};

const withConnect = connect(
  state => ({
    isLoading: fromState.isLoading(state),
    things: state.things.items,
    selectedThing: state.things.selectedItem,
    shouldShowNotFoundError: state.things.showNotFoundError,
  }),
  {
    getThings: thingActions.getThings,
    selectThing: thingActions.selectThing,
    showNotFoundError: thingActions.showThingNotFoundError,
  },
);

export default compose(
  withConnect,
  withRouter,
  withResetOnUnmount,
  injectIntl,
)(Things);
