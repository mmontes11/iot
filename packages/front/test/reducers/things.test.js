import deepFreeze from "deep-freeze";
import thingsReducer from "reducers/things";
import { THINGS_UPDATED, THING_SELECTED } from "constants/actionTypes/things";
import { RESET } from "constants/actionTypes/common";
import { initialState } from "../constants/index";

const thingsInitialState = initialState.things;
deepFreeze(thingsInitialState);

describe("reducers/things", () => {
  it("has a default state", () => {
    expect(thingsReducer(undefined, { type: "@init" })).toMatchSnapshot();
  });
  it("reduces nothing", () => {
    expect(thingsReducer(thingsInitialState, { type: "WHATEVER" })).toMatchSnapshot();
  });
  it("reduces THINGS_UPDATED", () => {
    expect(thingsReducer(thingsInitialState, { type: THINGS_UPDATED, things: ["foo", "bar"] })).toMatchSnapshot();
  });
  it("reduces THING_SELECTED", () => {
    expect(thingsReducer(thingsInitialState, { type: THING_SELECTED, thing: "foo" })).toMatchSnapshot();
  });
  it("selects and deselects a thing", () => {
    let state = thingsReducer(thingsInitialState, { type: THING_SELECTED, thing: "foo" });
    expect(state).toMatchSnapshot();
    state = thingsReducer(state, { type: THING_SELECTED, thing: "foo" });
    expect(state).toMatchSnapshot();
  });
  it("reduces RESET", () => {
    expect(thingsReducer(thingsInitialState, { type: RESET })).toMatchSnapshot();
  });
});
