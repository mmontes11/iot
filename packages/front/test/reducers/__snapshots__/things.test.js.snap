// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers/things has a default state 1`] = `
Object {
  "items": Array [],
  "selectedItem": null,
  "showNotFoundError": false,
}
`;

exports[`reducers/things reduces RESET 1`] = `
Object {
  "items": Array [],
  "selectedItem": null,
  "showNotFoundError": false,
}
`;

exports[`reducers/things reduces THING_SELECTED 1`] = `
Object {
  "items": Array [],
  "selectedItem": "foo",
  "showNotFoundError": false,
}
`;

exports[`reducers/things reduces THINGS_UPDATED 1`] = `
Object {
  "items": Array [
    "foo",
    "bar",
  ],
  "selectedItem": null,
  "showNotFoundError": false,
}
`;

exports[`reducers/things reduces nothing 1`] = `
Object {
  "items": Array [],
  "selectedItem": null,
  "showNotFoundError": false,
}
`;

exports[`reducers/things selects and deselects a thing 1`] = `
Object {
  "items": Array [],
  "selectedItem": "foo",
  "showNotFoundError": false,
}
`;

exports[`reducers/things selects and deselects a thing 2`] = `
Object {
  "items": Array [],
  "selectedItem": null,
  "showNotFoundError": false,
}
`;
